src/main/java/fi/fmi/avi/archiver/config/factory/postaction/SwimRabbitMQPublisherFactory.java

```diff
@@ -90,17 +104,30 @@ public SwimRabbitMQPublisher newInstance(final Config config) {
         final Config.ConnectionConfig connectionConfig = config.getConnection();
         final RabbitMQConnectionHealthIndicator healthIndicator = newHealthIndicator();
         final Environment environment = registerCloseable(newAmqpEnvironmentBuilder().build());
+
+        final AtomicReference<Connection> connectionRef = new AtomicReference<>();
+        final AtomicReference<Publisher> publisherRef = new AtomicReference<>();
+
         final Connection connection = createLazyForwardingProxy(Connection.class, () -> registerCloseable(environment
                 .connectionBuilder()
                 .username(connectionConfig.getUsername())
                 .password(connectionConfig.getPassword())
                 .uri(connectionConfig.getUri())
                 .listeners(SwimRabbitMQPublisherFactory::log, healthIndicator)
-                .build()));
+                .build()), connectionRef);
+
         final Publisher publisher = createLazyForwardingProxy(Publisher.class, () -> registerCloseable(connection.publisherBuilder()
                 .exchange(config.getExchange())
                 .key(config.getRoutingKey())
-                .build()));
+                .listeners(context -> {
+                    if (context.currentState() == Resource.State.CLOSED && context.failureCause() != null) {
+                        LOGGER.error("AMQP publisher closed unexpectedly - connection and publisher will be recreated on next publish attempt");
+                        publisherRef.set(null);
+                        connectionRef.set(null);
```

Comment:
To prevent infinitely storing obsolete references in memory, we should do `getAndSet` instead and then create a `unregisterCloseable` method to invoke with the retrieved non-null value. Should we also invoke `close()` just in case?

And now that elements could be removed from `closeableResources` in case of an error, I'd consider whether a `Set`/`HashSet` would be more efficient collection type. Though removal takes place only in an error case, so `List` might still be of a good choice.

Reply:
From the concurrency perspective, I'd guess we would like to unset the `publisher` as quickly as possible. Thus, first doing `getAndSet(null)` and then unregister and close. But I'd guess this would be just fine-tuning that doesn't make a difference in general. I'm happy with this.